File: 98/isValidBST.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
bool isValidBST(struct TreeNode* root) {
    bool status = true;
    struct TreeNode ** stack = (struct TreeNode**)malloc(sizeof(struct TreeNode*) * 50);
    int length;
    struct TreeNode * node;

    if(!root) return true;

    if(root->left){
        length = 0;
        stack[length++] = root->left;
        while(length > 0){
            node = stack[--length];
            if(node->val >= root->val) return false;

            if(node->right) stack[length++] = node->right;
            if(node->left) stack[length++] = node->left;
        }
    }

    if(root->right){
        length = 0;
        stack[length++] = root->right;
        while(length > 0){
            node = stack[--length];
            if(node->val <= root->val) return false;

            if(node->right) stack[length++] = node->right;
            if(node->left) stack[length++] = node->left;
        }
    }

    return (status && isValidBST(root->left) && isValidBST(root->right));
}

File: 98/isValidBST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True

        return self.isValid(root,-(1<<32),(1<<32))

    def isValid(self,root,small,large):
        if not root:
            return True

        if root.val > small and root.val < large:
            return self.isValid(root.left,small,root.val) and self.isValid(root.right,root.val,large)
        else:
            return False
